### ISSUE TITLE
Untitled story 6478

### ISSUE THREAD

**Issue by diegoliberman** [instedd/cdx#1485]

**Comment by straight-shoota** [comment]
@diegoliberman The current content of the OP looks like mostly a duplicate of  #1503. IIRC this originally had a different, more general scope. Right?
The edit history seems unreliable. The change shown there doesn't really make sense to me.

**Comment by diegoliberman** [comment]
@straight-shoota sorry. It looks like I mixed the tickets. This one is for doing an overall assessment.

**Comment by devduarte** [comment]
I think that the best (and correct) way to go here is to find a way to add a nested form for the `AssayAttachment` model. We should be able to add the nested form dinamically, so we can keep the current UX. If we manage to accomplish this, it will save us a lot of JS code that we are using to mimic how rails handles html input names and other kind of stuff. It will also allow to use the correct relation between `AssayFiles` and `AssayAttachments`, since currently there is an `after_save` method that works as a workaround for not having the nested form and the correct relation between these two models.

**Comment by straight-shoota** [comment]
I had already considered [vanilla_nested](https://github.com/arielj/vanilla-nested/) for this. It's a simple shard with no dependencies. Just two view helpers and a vanilla JS-implementation. It takes care of most of the custom JavaScript we're currently using for the assays form.

**Comment by straight-shoota** [comment]
I think there can be many simplifications to the implementation of assay attachments.

Currently, most of the implementation is based around `dropzonejs`. We can continue using `dropzonejs` as a convenience enhancement, but it should not be the core that everything else builds upon.
We can drop most of the custom JavaScript in https://github.com/instedd/cdx/blob/933a350c83bb8f5e2cdf7eda8b5e8e70dcdd38af/app/views/samples/_form_assays_js.haml and replace it by `vanilla_nested`.
The nested forms for assay attachment should first be implemented with only basic rails form features. It can be gradually enhanced for more advanced functionality to achieve the desired UI of #1465.

Before starting that, I would add specs for the current behaviour. The user interaction is mostly fine, we just want to refactor the implementation. To make sure we avoid regressions, it makes sense to have tests. Currently, there are no integration specs for the samples form.

`vanilla_nested` requires Rails 4.2.11, so #1511 is a perfect base for that. I already have a simple PoC.

**Comment by straight-shoota** [comment]
I'm wondering about a few details in the current implementation:

1. `dropzonejs` has a custom handler for client-side file name sanitization. I think this kind of stuff should better happen on the server. I'm also wondering about the specifics of this action. Besides sanitizing non-alphanumeric characters it also adds a timestamp to the file name. Is that a functional requirement or just a measure to avoid collisions?
2. The *LOINC* field is a a plain text field with an attached autocomplete widget. That's different from other auto-complete fields such as *Specimen Role* and results in a different UI. I'm not sure what's the reason for using a different approach to the established UI component for autocomplete fields.
3. Just to verify: Is the ability to drag and drop upload (multiple) files as assay attachments a functional requirement? Should drag and drop upload also work for the file field of individual attachments? (cf. https://github.com/instedd/cdx/issues/1465)

**Comment by straight-shoota** [comment]
A similar simplification for the nested form can be done for *Notes* as well.

**Comment by diegoliberman** [comment]
@straight-shoota great progress! The feature of uploading multiple files isn't a critical requirement.  I think it was a product of our own initiative.  Clearly, it's a nice feature when you need to upload multiple files.  But I wouldn't know how often that would be needed.  

I think there was no concrete requirement in regards to usability.  It was just a matter of allowing to add Assays, with its corresponding attachment, result, and LOINC code.  As long as we cover that, we are fulfilling the user's need.  So, if you tell me that removing the possibility of doing drag & drop (for instance), simplifies significantly the implementation, we should absolutely discuss it.

**Comment by devduarte** [comment]
> I'm wondering about a few details in the current implementation:
> 
> 1. `dropzonejs` has a custom handler for client-side file name sanitization. I think this kind of stuff should better happen on the server. I'm also wondering about the specifics of this action. Besides sanitizing non-alphanumeric characters it also adds a timestamp to the file name. Is that a functional requirement or just a measure to avoid collisions?
> 2. The _LOINC_ field is a a plain text field with an attached autocomplete widget. That's different from other auto-complete fields such as _Specimen Role_ and results in a different UI. I'm not sure what's the reason for using a different approach to the established UI component for autocomplete fields.
> 3. Just to verify: Is the ability to drag and drop upload (multiple) files as assay attachments a functional requirement? Should drag and drop upload also work for the file field of individual attachments? (cf. [Improve UX/UI in Assays #1465](https://github.com/instedd/cdx/issues/1465))

1) In think that the main reason we added a timestamp into the  filename, was because filenames with a single special character (like "\\") caused problems when being read by rails before loading the AssayAttachment card.

2) Specimen_Role uses a react component that has all of its options loaded in memory. Loinc Code input uses another autocomplete component because its a table with more than 90k records, this autocomplete allow us to do a search via ajax without the need of loading too many options in memory.

3) Designs show a "Drop assays here" in the AssayAttachment sections, so we build it with a drag and drop feature. Overall it was one of the least time consuming behaviour, so it didn't represent of effort to have it.

> I think there can be many simplifications to the implementation of assay attachments.
> 
> Currently, most of the implementation is based around `dropzonejs`. We can continue using `dropzonejs` as a convenience enhancement, but it should not be the core that everything else builds upon. And maybe we can find an alternative that's easier to integrate. There are options for `ActiveStorage` which we should probably migrate to at some point.
> 
> We can drop most of the custom JavaScript in https://github.com/instedd/cdx/blob/933a350c83bb8f5e2cdf7eda8b5e8e70dcdd38af/app/views/samples/_form_assays_js.haml and replace it by `vanilla_nested`. The nested forms for assay attachment should first be implemented with only basic rails form features. It can be gradually enhanced for more advanced functionality to achieve the desired UI of #1465.
> 
> Before starting that, I would add specs for the current behaviour. The user interaction is mostly fine, we just want to refactor the implementation. To make sure we avoid regressions, it makes sense to have tests. Currently, there are no integration specs for the samples form. This is tracked in #1468.
> 
> `vanilla_nested` requires Rails 4.2.11, so #1511 is a perfect base for that. I already have a simple PoC.


Migrating to ActiveStorage was always in our plans, but we needed to upgrade rails first. Seems like a good time to evaluate it again to check if its feasible now!
Any option that would allow us to get rid of 'manually' handling form inputs is very welcome, we didn't feel super comfortable doing that and didn't know about gems for dynamically adding nested forms. Vanilla_nested seems what we need!

**Comment by straight-shoota** [comment]
> 2. Specimen_Role uses a react component that has all of its options loaded in memory. Loinc Code input uses another autocomplete component because its a table with more than 90k records, this autocomplete allow us to do a search via ajax without the need of loading too many options in memory.

From a UX perspective, this seems more like a job to enhance the existing select component for autoload support, than creating a new component for this.

The `react-select` component we're using seems to be an old version of https://react-select.com/. At least newer versions of that definitely support autoloading. But even the version in our codebase has an `asyncOptions` property which seems to serve exactly this purpose. I'll try to use that.

**Comment by ysbaddaden** [comment]
Oh, `vanilla_nested` looks very nice!

ActiveStorage would have helped so much to implement this feature that it hurts. We definitely must target it with the refactor: we could remove all the custom code, and even allow some nice enhancements, such as direct uploads to S3.

Let's not fret over the file drop: it's ubiquitous nowadays, and very easy to implement, even with vanilla JS/DOM. It's basically a DOM attribute and an event listener.

**Comment by diegoliberman** [comment]
@straight-shoota I don't think we will be able to refactor this year.  But perhaps it's a good idea to write down the plan that we had, now that Rails 5 is in Production.  Could you add some bullets to help ourselves in the future?

**Comment by straight-shoota** [comment]
The battle plan is:

* Refactor the client side UI to use a form helper for nested attributes (recommend `vanilla_nested`). This also simplifies some of the server side code.
* refactor file uploads to use `ActiveStorage`
* apply new UI design
* Use CdxSelect with autoloading for loinc code
